### DISCUSSION TITLE
What is your usage of the `--max_requests` option?

### DISCUSSION THREAD

**Discussion by benoitc** [benoitc/gunicorn#3042] [General]
Serious question but why so much people are using  the option `--max_requests` run run gunicorn?

 This option has been added to Funicorn to offer  a temporary workaround for an application code leaking memory. But it seems people are using it for another usage.

Please let me know. There maybe a better way than forcing gunicorn to kill that worker .

**Comment by sillycube** [comment]
For my 2 vCPU machine, I use this option as 
```
gunicorn xxxx.wsgi:application --workers=5 --threads=5 --max-requests=1000
      --max-requests-jitter=30 --worker-class=gthread -b 0.0.0.0:8000
```

I use it in my Django app because memory usage is increasing after a while. So I want to rectify this issue automatically.

**Comment by transreductionist** [comment]
@benoitc We are using 4 workers and 2 maximum requests in our primary environment. In another we use 1 worker and 0 maximum requests to accommodate certain testing I am told. After some analysis today we are moving to N workers and 0 maximum requests, which is the point of this discussion. Our memory usage remains stable across multiple requests, and memory leaks appear not to be an issue.

Perhaps you will have some time to comment on the following. It would be appreciated.

### Why Use Maximum Requests

Why do we use this configuration? The configuration was chosen years ago, and it is not clear today why 4/2 was chosen. It is being revisited today because of some infrastructure and environment changes. From program history it seems like there were memory issues at some point in our containers, and it was thought that this might be due to memory leaks. The choice was made to reboot the workers using maximum requests.

After some analysis today we are moving to N workers and 0 maximum requests.

### A Current Example

One of the tests that was run recently to gain a better understanding of what happens when a worker is rebooted uses one worker and N=4 maximum requests. This configuration was chosen for its simplicity, and the ability to get to the maximum requests quickly. Also, it was easy to monitor the gunicorn worker and watch it restart with a different PID. Here is what was observed. At the start of the gunicorn server:

```shell
Starting gunicorn with 1 workers and 4 max-requests
2024-12-04 13:48:17 [8] [gunicorn.error:278] [INFO] Starting gunicorn 22.0.0
2024-12-04 13:48:17 [8] [gunicorn.error:278] [INFO] Listening at: http://0.0.0.0:6327 (8)
2024-12-04 13:48:17 [8] [gunicorn.error:278] [INFO] Using worker: sync
```
- The first request is sent and is handled end-to-end: N=1.
- Next, 2 near simultaneous requests are made, and they are both handled end-to-end: N=3.
- On the next request where N=4=maximum request:
```shell
2024-12-04 13:57:55 [gunicorn.error:278] [INFO] Autorestarting worker after current request.
... The application continues logging.
2024-12-04 13:57:59 [ERROR] traceback [File "/opt/miniconda3/lib/python3.11/concurrent/futures/thread.py", line 169, in submit;
        raise RuntimeError('cannot schedule new futures after')]
2024-12-04 13:58:00 [144] [gunicorn.error:278] [INFO] Booting worker with pid: 144
```  
This behavior is similar to reference (1) below, but there they have 2 workers and suggest that as the 1st worker reaches its maximum requests the second worker, which is handling a request, submits a task to ThreadPoolExecutor and a failure occurs: `cannot schedule new futures after`. 

### Observation 

What I see happening here is that as the request comes in (N=4), gunicorn signals it is autorestarting the worker at 13:57:55. Five seconds later gunicorn signals that is is booting the new worker. Somewhere within those 5 seconds an error is thrown. That makes sense to me because the worker that was handling the request is no longer alive to handle the request. 

Whether there is 1 worker or many workers, anytime a worker reaches its max requests, this sort of behavior will exhibit itself because the worker handling that particular request is no longer alive.

### Two Questions

- Is this assessment accurate, or have I missed something that lays deeper?  
- If a given worker is restarted does it affect the process of a different worker?  

### Some Related Issues

1. [A request raises RuntimeError when submitting ThreadPoolExecutor future on gunicorn with --threads and max_requests options](https://github.com/benoitc/gunicorn/issues/3007)  
2. [Connection reset during max-requests auto-restart with gthread](https://github.com/benoitc/gunicorn/issues/3038)  
3. [Prevent gthread connection reset on max-requests restart](https://github.com/benoitc/gunicorn/pull/3039)

**Comment by klausfrandsen** [comment]
We're using gunicorn on Heroku to serve a Django app.

Note that Heroku mentions --max-requests as a "convenient way to help limit the effects of the memory leak":
https://devcenter.heroku.com/articles/python-gunicorn

So that might a reason for many folks to use the setting.

After upgrading to 21.x we've also run into the #3340 issue and friends. I will experiment with turning restarts off (--max-requests=0) until a proper fix hopefully will be available.

**Comment by benoitc** [comment]
just don't use ma request if you don't need it ...

**Reply by klausfrandsen** [reply]
True, we've had it set for a loong time (about three years). Back then we enabled it to try and limit our memory footprint after adding some heavy monitoring tools. And then it never got removed again, after that monitoring stuff was reverted. And it worked fine until a very recent round of upgrades.

But whether you need it or not is not a clear-cut answer. Complex applications often leak to some extend, and often it happens in 3rd party libs that are not easy to fix. Maybe your traffic is low enough in the beginning, that you don't notice small leaks. On Heroku this is especially true, as they enforce daily restarts of all dynos. Later on, leaks might build up faster under more heavy traffic. And that has to be compared to your RAM limit on the Heroku dyno. 

I don't pretend to understand the details discussed in #3340, and I fully acknowledge that it is tricky to fix. But until a fix is available, maybe mention this in some form or another in the docs.

And thanks a lot for a great package. Couldn't do without it :-)

**Reply by klausfrandsen** [reply]
Just as a followup, the memory footprint of Our application has increased somewhat after we turned max-requests off. See the picture below, where the cursor indicates when it was turned off.

![image](https://github.com/user-attachments/assets/7b73a482-a5a0-4a1c-9a92-db74c1b018d2)

**Reply by Meess** [reply]
We're encountering the same issues and use it for the same purpose, using --max-requests because memory is slowly ramping up (we're suspecting it's on of the add-on's of heroku, but a haven't got the time to pinpoint which), and mainly occurs when we add the --threads parameter. (and we're using the --thread parameters to support more long lived requests with Django, for e.g. LLM streamed in responses. Heroku has very low memory so adding threads is way more efficient then adding workers). We're also experiencing the occasional connection interrupted as described in https://github.com/benoitc/gunicorn/pull/3340

**Comment by dbartenstein** [comment]
@benoitc we have also been using Gunicorn for many years now together with the latest Django LTS versions.

We had `--max-requests` set to 50_000, which ended up being too high. This led to the kernel’s OOM-killer kill the Gunicorn process (red arrow in the screenshot below). As a mitigation, we now have reduced `--max-requests` to 10_000 and the number of workers from 5 to 4.
Additionally we’ll experiment with systemd’s `MemoryHigh` and `MemoryMax` options so see if we can contain Gunicorn within memory bounds.

As you can see, without the automatic worker restart, we’d run out of memory very quickly:
<img width="1346" height="295" alt="image" src="https://github.com/user-attachments/assets/2d26d78b-20ef-4671-92fe-6aef842cb65e" />

Let me also share Adam’s very interesting blog entries with you:
- https://adamj.eu/tech/2021/12/29/set-up-a-gunicorn-configuration-file-and-test-it/
- https://adamj.eu/tech/2019/09/19/working-around-memory-leaks-in-your-django-app/

Recommendations by service providers:
- https://www.digitalocean.com/community/tutorials/how-to-set-up-django-with-postgres-nginx-and-gunicorn-on-ubuntu#advanced-performance-optimizations
